drafts/angular-navigation-concept.md

```diff
@@ -0,0 +1,607 @@
+We cannot intercept the native browser back button, meaning the user will always end up on the empty start page.
+However, for the `back` button within our application, we can implement a fallback if we detect this scenario.
+The fallback is to replace the current page with the home page.
+
+## Prevent Exiting The Application
+
+To determine if a back navigation would leave the application, we first have to know how large the history stack is.
+Instead of tracking the size manually, we can use an internal `Angular Router` counter, which is persisted in the history state.
+The `ɵrouterPageId` gets updated by the router before the guard checks are run.
+This means if we access the `pageId` in a `canActivate` guard, it will already be updated.
+
+However, if you know a little about Angular, you might know that functions and properties beginning with a Greek theta (`ɵ`) are usually private to Angular internals and should not be used.
+Angular declares these functions and properties as unstable since they are not part of the public API.
+However, in the Angular source code, they added this comment:
+
+```ts
+// The `ɵ` prefix is there to reduce the chance of colliding with any existing user properties on
+// the history state.
+```
+
+This does not guarantee that this property will not change in future updates.
+_If it is removed, it is not that difficult to build this yourself._
+
+To make this work, we only have to extend the `HistoryState` model and update the `NavigationService.historyGo` function.
+
+```ts
+export interface HistoryState {
+  ɵrouterPageId?: number; // <---- Add
+  flowSourcePageOffset?: number;
+  mainPageOffset?: number;
+}
+```
+
+Once we have access to the `ɵrouterPageId` state, we simply check if the current navigation request would exit the application.
+If it does, we replace the current page with the home page.
+
+```ts
+
+@Injectable({ providedIn: 'root' })
+export class NavigationService {
+  private readonly locationService = inject(LocationService);
+  private readonly router = inject(Router);
+
+  back(): void {
+    this.historyGo(-1);
+  }
+
+  private historyGo(relativePosition: number) {
+    const pageId = this.locationService.getCurrentState()?.ɵrouterPageId ?? 0;
+    if (pageId + relativePosition < 0) {
+      this.router.navigate(['/home'], { replaceUrl: true });
+      return;
+    }
+    this.locationService.go(relativePosition);
+  }
+}
+```
+
+# Calculating The Offsets
+
+If we recall the concept, we want to store an offset to the `flow source page` and `main page` in the history state.
+The calculation of the offset should happen once a navigation finishes.
+The `Angular Router` provides an observable for navigation events.
+We can subscribe to `router.events` and listen for `NavigationEnd`.
+At this point, we know the navigation has successfully finished, and we can update the state.
+One important note is that we do not want to increase the offset if we navigate using `{ replaceUrl: true }`.
```

Comment:
Idea: Explain why is `replaceUrl` used, what does it do, what are the implications.